### ISSUE TITLE
Duplicate orders with same Quote Id at same time with few time difference.

### ISSUE THREAD

**Issue by amitvishvakarma** [magento/magento2#13952]
**Preconditions**

1. Ability to process requests at same time. (in our case, we had use 2 different Client's system.)

**Steps to reproduce**
    

1. Login as a User
1. Add items to cart and proceed through checkout.
1. Select Any payment method. Not click on place order button
1. Now login to the same account in another system.
1. And proceed to checkout with the same items. Not click on place order button.
1. Now you have to click on place order at the same time with both the systems.
Observe order confirmation screen of both systems. **(usually first time you will get an error about constraint violations. Try to repeat steps 1-6 again. Issue reproduces 100% from the second attempt)**

Depending on your infrastructure, this bug may be challenging to reproduce. It requires all of the "Place order" button clicks to be processed concurrently by Magento. You will not see the bug if Magento is processing requests sequentially (as the second request will reference a cart ID that no longer exists).

**Expected result**
A single order has been placed.
Alternatively, if multiple requests are unavoidable: Only the first request to be processed.

**Actual result**
Two different orders have been placed with very few seconds as time difference with same quote Id.

**Comment by magento-engcom-team** [comment]
@amitcedcoss, thank you for your report.
We've acknowledged the issue and added to our backlog.

**Comment by sunilit42** [comment]
I am working on this at #dmcdindia

**Comment by sunilit42** [comment]
It happens when you click same time on the button.
@magento-engcom-team  There is two way we can do
1) use flock, for that we need to create one lock file quote wise when you click on that button
2) Either create one table and make an entry while click on submit and remove when order placement or fail order.

Let me know which way you want to process.

**Comment by timbaker1991** [comment]
Just to further this, we have the issue and it appears to affect PayPal orders mainly- though this may be coincidence (using the baked in Braintree support this is). Is there anything we could use as a hot fix do you know?

**Comment by amitvishvakarma** [comment]
@magento-engcom-team any update regarding our issue?

**Comment by buituanbg** [comment]
@magento-engcom-team I got the same issue too. The quote is same ID but it created 5 orders today. So I appreciate any updates for this.

**Comment by tzy1992** [comment]
@magento-engcom-team Any update on this issue? The issue occur randomly and it created multiple order on my site mainly from using Cash on Delivery payment method.

**Comment by buituanbg** [comment]
@tzy1992 I had a quick fixing solution. let me know if you want to view that.

**Comment by tzy1992** [comment]
@buituanbg Please share the fix with me. Thank you

**Comment by timbaker1991** [comment]
@buituanbg Please could you share it on here for us all to see if possible.

**Comment by buituanbg** [comment]
Hi @tzy1992  and @timbaker1991 
I created an observer as below:
app/code/TB/FixDuplicateOrder/Observer/OrderPlacebefore.php
'<?php
namespace TB\FixDuplicateOrder\Observer;

use Magento\Framework\Event\Observer; 
use Magento\Framework\Event\ObserverInterface;

class OrderPlacebefore implements ObserverInterface
{
	protected $_responseFactory;
    protected $_url;

    public function __construct(
        \Magento\Framework\App\ResponseFactory $responseFactory,
        \Magento\Framework\UrlInterface $url
    ) {
        $this->_responseFactory = $responseFactory;
        $this->_url = $url;
    }

	public function execute(\Magento\Framework\Event\Observer $observer)
	{
		$event = $observer->getEvent();

		$order = $observer->getEvent()->getOrder();
		$quoteID = $order->getQuoteId();
		$objectManager = \Magento\Framework\App\ObjectManager::getInstance();  
		$resource = $objectManager->get('Magento\Framework\App\ResourceConnection');
        $connection = $resource->getConnection();
        $tableName = $resource->getTableName('sales_order');
        $selectedTime = date('Y-m-d h:i:s');
		$endTime = strtotime("-15 seconds", strtotime($selectedTime));
		$last15Sec = date('Y-m-d h:i:s', $endTime);

        echo $sql = "SELECT * FROM `".$tableName."` WHERE `quote_id` = ".$quoteID." and `created_at` >= '$last15Sec'";
        if($result = $connection->fetchRow($sql)){
        	$customerBeforeAuthUrl = $this->_url->getUrl('checkout');
	        $this->_responseFactory->create()->setRedirect($customerBeforeAuthUrl)->sendResponse();
	        exit();
        }
        //var_dump($result);
	    //echo  "Hello Testing ";
	    //die();
	}
}'
 app/code/TB/FixDuplicateOrder/etc/events.xml 
<?xml version="1.0"?>
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"   xsi:noNamespaceSchemaLocation="urn:magento:framework:Event/etc/events.xsd">
<event name="sales_order_place_before">
    <observer name="sales_order_load_tip_and_trick_attribute" instance="TB\FixDuplicateOrder\Observer\OrderPlacebefore" />
</event>
</config>

app/code/TB/FixDuplicateOrder/registration.php
'<?php
 
\Magento\Framework\Component\ComponentRegistrar::register(
    \Magento\Framework\Component\ComponentRegistrar::MODULE,
    'TB_FixDuplicateOrder',
    __DIR__
);'
This event observer will prevent order created on the same quote in past 15 secrets because as my detect main issue in about 15 secret it create order with same quote id. I created this from 01-Jul and I didn't get any duplicate order after that. Also all other process relate to place order is fine there isn't any issues. 
Hope my solution can help. Let me know if any.
Thanks,
tuan

**Comment by tzy1992** [comment]
@buituanbg Thanks for sharing the fix, for the event you are using `sales_order_place_before` ?

**Comment by buituanbg** [comment]
yes, I used sales_order_place_before event.

**Comment by tzy1992** [comment]
@buituanbg I am still getting duplicated order after apply the fix, both order created at the exact date and time.

@magento-engcom-team Any update regarding this issue?

**Comment by buituanbg** [comment]
@tzy1992 it seems you got other case. Let me know if you need support to debug.

**Comment by misha-kotov** [comment]
@tzy1992 @buituanbg @timbaker1991 @amitcedcoss 
Does anyone have a guess why this happens on production systems? I doubt real shoppers are checking out in different browsers/devices trying to cause race conditions on your stores. We have heard a theory that when the system is overloaded, order submission may be kicking users back into checkout even if the order is created, and submitting again creates a duplicate order. Currently the quote does get invalidated, but there is still a window of opportunity of a couple seconds where duplicate orders can get created. 
Just trying to get to root cause so we can fix appropriately.

**Comment by rauberdaniel** [comment]
General question: What would happen if there would be a `UNIQUE` constraint set to the `quote_id` column of the `sales_order` table. That would prohibit the creation of duplicate orders for the same quote in any case, however, I’m not sure if it has any problematic side effects or what the result on customer side would be in such a case.

(But in general I feel like this constraint should be set anyway?)

**Comment by DependencyHell** [comment]
@rauberdaniel I think the `quote_id` cannot be `UNIQUE` as if one order is canceled (getting back from paypal without validation for example), the cart (`quote_id`) remains the same and can be used in another order.

**Comment by developer-vtnetzwelt** [comment]
Any update on this issue?

**Comment by mgomma** [comment]
Any update on that issue .. it happens when create order from custom code .. not happen when place order regularly.

**Comment by LSERRE** [comment]
This issue occurs randomly on one of our 2.3.0 project. Any update on this issue?

**Comment by oxvijay** [comment]
We are also getting the duplicate orders for the same quote in few seconds interval. The issue is not consistent and can't be reproduced every time. But it's an obvious issue we are facing frequently. Please help if you know any workaround. The issue was in 2.2.x and we are in 2.3.0 now.

**Comment by TonyRutherford1** [comment]
Also having this issue with 2.3.0 randomly. Any fix or workaround?

**Comment by TonyRutherford1** [comment]
@magento-engcom-team Have upgraded to 2.3.1 and still geting this issue randomly with orders being duplicated. Is anyone looking at this?

**Comment by david-fiaty-cko** [comment]
+1

**Comment by bjclark13** [comment]
@magento-engcom-team any updates on this year+ old issue?

**Comment by dtfdigital** [comment]
A client of ours has experienced this on 2.3.2 with a customer paying via paypal. The second order was created about 30 seconds after the first, and was missing the paypal transaction ID.

**Comment by hostep** [comment]
Hmm, a client of ours recently reported something similar to this, not sure if this is the exact same problem as being discussed over here, but it kind of sounds similar, steps to reproduce:

- Put products in your cart in the shop
- Go through the checkout and pick Paypal
- Be redirected to the Paypal
- Close your browser
- Open your browser
- Visit the same shop again
- Notice that your cart is still there
- Attempt to go through the checkout and pick Paypal again
- A crash occurs with the message 'We can't initialize Express Checkout' (from memory: I think this was only being logged and not shown to the customer)

This was on Magento 2.1.16, I have no idea if something has changed in newer versions of Magento around this.

@misha-kotov: is this a valid scenario? A client of ours sometimes complains about it but it doesn't seem to happen that often in practice.

If this isn't the same bug as the thread, please ignore this 🙂

**Comment by fmjrey** [comment]
Hi, I'm no Magento developer, just a knowledgeable user of a web site that uses Magento and I've seen this this bug happening. Basically my paypal payment on that web site resulted in two orders created on their backend. I prefer not to give the details of the web site, but I can confirm this bug also happens from a single successful payment (I can see the payment on my bank app), no need for a concurrent session in another browser.

In my case it happened as follows: after I submitted the payment details in the paypal window, I saw it close and at about the same time I moved the mouse too close to the keyboard.
When the mouse hit the keyboard its back button was inadvertently pressed on the seller page.
I suspect this sent a second "payment" query to the server which went on to create a second order based on the same details.
It is likely that this second query still contain some form of payment/transaction/token id, as this is all happening from the same browser.

I can't be more specific here as  I'm only a user of the website. I will however point them to this issue here, they're the one who can continue the discussion.

**Comment by Ashna-Jahan** [comment]
Hi @magento-engcom-team,

We are still facing the above issue in version 2.3.2, we are able to place multiple orders at the same time when clicked simultaneously from 2 different browsers.

We found a solution, wherein we could make the quote_id field in the sales_order table as a UNIQUE field. Kindly let us know if this could be right approach to follow.

**Comment by dufrygrant** [comment]
> Hi @magento-engcom-team,
> 
> We are still facing the above issue in version 2.3.2, we are able to place multiple orders at the same time when clicked simultaneously from 2 different browsers.
> 
> We found a solution, wherein we could make the quote_id field in the sales_order table as a UNIQUE field. Kindly let us know if this could be right approach to follow.

Did you find that this ended up being a durable solution to the problem?

**Comment by joshiprashiddha** [comment]
Hi @sdzhepa , where is the solution ? Why you added ready of QA ?

**Comment by sdzhepa** [comment]
Hello @joshiprashiddha 

Unfortunately, I do not have a solution for this issue.
The label `Progress: ready for QA` means this issue should be re-verified on the lates `2.4-develop` according to [Issues Processing Workflow](https://github.com/magento/magento2/wiki/GitHub-Issues-Processing-Workflow).
When the report will be verified and receive label `Issue: Confirmed`, label `Progress: ready for dev ` will be set.
Or if any contributor creates Pull Request to fix it, label "Progress: PR in progress" will be set.

All this action is reflecting on the [ Issue Board: Community Backlog](https://github.com/magento/magento2/projects/20)

**Comment by AnshuMishra17** [comment]
I am facing the same issue in 2.3.4. I can see duplicate orders from the same quote id but I am not able to reproduce it myself.

**Comment by joshiprashiddha** [comment]
This happens once in a while. What I can guess from accumulated data upto now is, its very very hard to reproduce. And its happening due to reserve order id.

**Comment by trunglv** [comment]
I think it's not an issue of Magento itself. It relates to several modules that work on a checkout process are DHL, Paypal, Amazon Pay ... I fixed the issue for one client. Some small fixings I found on the internet just to prevent Magento to make an order again for already existed quote id ... but don't solve real problems. Just my idea to solve that issue 
- It's really hard to reproduce it on TEST/STAGING server - often happen on a LIVE server.
- Pls check your checkout - whether place order button can be clicked double or many times 
- Pls check third-party shipping methods - Do they cause any problem that stops a checkout process/ a submit order process? - Check logs or you can write a log to monitor it. 
- Check third-party payments as well

**Comment by jpratt** [comment]
Seeing the same on our production on 2.3.5. Our orders are a minute or two apart with the same quote id. Using braintree for cc payments. Very random, and see the second order does not send a notification, that is the only difference I can see.

**Comment by helloeverything** [comment]
Seeing this on 2.3.5-p1 using stripe, multiple orders a minute or two apart with same transaction ID.

**Comment by bakennguyen** [comment]
Hi, I can also confirm it is reproduced in 2.3.4-p2

### steps to reproduce:

- add items in your cart
- go to cart
- checkout with paypal
- go to review page
- click place Order, then quickly cancel page submission
- refresh review page

### observed behavior
an order is created with email notification
the review page is still available and user can place another order

### expected behavior
User should be redirected to success page of the previously created order

**Comment by sarjangautam** [comment]
Hi,
We are getting this issue in our production with Magento2 Enterprise 233. After investigating the issue we were able to replicate this in our local and this is happening in below scenarios:
1. Customer clicks place order order button in checkout page. 
2. Payment is accepted/taken by payment providers (Cybersource, Paypal etc)
3. Once response is received from payment provider, magento quote management class places order or creates order in DB. 
4. Once order is created next step is saving the quote. If there is any error/exception occur during saving quote then customer is sent to My Cart page with message saying there was some issue creating order, try again etc .This stage customer does not know if order is placed or not in DB and they must probably try again to checkout resulting another order for same item/same quote

The class and method that is causing issue in our case is :
`src/vendor/magento/module-quote/Model/QuoteManagement.php::submitQuote`

This class has following try catch block at the end:
`
     try {
            $order = $this->orderManagement->place($order);
            $quote->setIsActive(false);
            $this->eventManager->dispatch(
                'sales_model_service_quote_submit_success',
                [
                    'order' => $order,
                    'quote' => $quote
                ]
            );
            $this->quoteRepository->save($quote);
        } catch (\Exception $e) {
            $this->rollbackAddresses($quote, $order, $e);
            throw $e;
        }`

We can see that order is placed first and then quote is saved. In our case the quote gets changed and some data seems to be invalid causing the logic to go in exception. Which results customer to sent back to My cart page with error message. However, the order is already placed in DB and customer will not know about this .

As a solution we have done following:
`       
     //place the order
        try {
            $order = $this->orderManagement->place($order);
        } catch (\Exception $e) {
            $this->rollbackAddresses($quote, $order, $e);
            throw $e;
        }

      //save the quote
        try {
            $quote->setIsActive(false);
            $this->eventManager->dispatch(
                'sales_model_service_quote_submit_success',
                [
                    'order' => $order,
                    'quote' => $quote
                ]
            );
            $this->quoteRepository->save($quote);
        } catch (\Exception $e) {
            $backupQuote->setIsActive(false);
            $this->quoteRepository->save($backupQuote);
        }`

Here `$backupQuote` is clone of original `$quote` object and this variable is created at the start of the method as below:
`$backupQuote = clone $quote;`
with this change customer  will be redirected to success page. Although the quote might not be accurate with this approach but we might not need the old quote once order is placed and we have marked the quote as inactive.

**Comment by arunkumar247** [comment]
We are facing this issue in 2.3.5-p2 version. Please update us if any fix available for this issue.

**Comment by dkilcoyne** [comment]
> We are facing this issue in 2.3.5-p2 version. Please update us if any fix available for this issue.

- Have you made any traction? We're having the same issue as well. 2.3.5-p1. We're working with Magento to confirm and patch the issue but it's been slow going.

**Comment by dkilcoyne** [comment]
Seems to be related to a db lock updating the # of times a promo code was used. When this times-out the user is given error where they order again, with the same quoteId. You can see the same rule_id being locked.
![log1](https://user-images.githubusercontent.com/580036/93230565-cc437b00-f745-11ea-9944-fbfd19ede6d0.jpg)
![log2](https://user-images.githubusercontent.com/580036/93230569-ccdc1180-f745-11ea-8a94-24009d9881c0.jpg)
![log3](https://user-images.githubusercontent.com/580036/93230570-ccdc1180-f745-11ea-9bf3-43933e07f105.jpg)

**Comment by Jazzz09** [comment]
> We are facing this issue in 2.3.5-p2 version. Please update us if any fix available for this issue.

I was also facing the same problem in 2.3.5-p2. In my case the quote was still active even after order was placed so I fixed this by observer for sales_order_place_before:

**events.xml:**
`<?xml version="1.0"?>
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Event/etc/events.xsd">
	<event name="sales_order_place_before">
		<observer instance="Vendor\Module\Observer\OrderPlaceBefore" name="duplicateorder_observer_orderplacebefore"/>
	</event>    
</config>`

**Vendor\Module\Observer\OrderPlaceBefore**
`<?php

namespace Vendor\Module\Observer;

class OrderPlaceBefore implements \Magento\Framework\Event\ObserverInterface
{
    /**
     * @var \Magento\Sales\Model\ResourceModel\Order\CollectionFactory
     */
    protected $orderCollectionFactory;

    /**
     * @var \Magento\Framework\App\ResponseFactory
     */
    protected $responseFactory;
    
    /**
     * @var \Magento\Framework\UrlInterface
     */
    protected $url;
    
    /**
     * @var \Magento\Framework\Stdlib\DateTime\DateTime
     */
    protected $date;
    
    /**
     * @var \Psr\Log\LoggerInterface
     */
    protected $logger;

   /**
     * Constructor function
     *
     * @param \Magento\Sales\Model\ResourceModel\Order\CollectionFactory $orderCollectionFactory
     * @param \Magento\Framework\App\ResponseFactory $responseFactory
     * @param \Magento\Framework\UrlInterface $url
     * @param \Magento\Framework\Stdlib\DateTime\DateTime $date
     * @param \Psr\Log\LoggerInterface $logger
     * @param \Magento\Quote\Model\QuoteFactory $quoteFactory
     */
    public function __construct(
        \Magento\Sales\Model\ResourceModel\Order\CollectionFactory $orderCollectionFactory,
        \Magento\Framework\App\ResponseFactory $responseFactory,
        \Magento\Framework\UrlInterface $url,
        \Psr\Log\LoggerInterface $logger,
        \Magento\Quote\Model\QuoteFactory $quoteFactory
    ) {
        $this->orderCollectionFactory = $orderCollectionFactory;
        $this->responseFactory = $responseFactory;
        $this->url = $url;
        $this->logger = $logger;
        $this->quoteFactory = $quoteFactory;
    }


    /**
     * Execute observer
     *
     * @param \Magento\Framework\Event\Observer $observer
     * @return void
     */
    public function execute(
        \Magento\Framework\Event\Observer $observer
    ) {
        try {
            $order = $observer->getEvent()->getOrder();
            $quoteID = $order->getQuoteId();
            $ordersCollection = $this->orderCollectionFactory->create();
            $ordersCollection->addFieldToSelect('quote_id');
            $ordersCollection->addFieldToFilter('quote_id', $quoteID);
            
            if ($ordersCollection->getSize() > 0) {
                //Load quote by quote id
                $quote = $this->quoteFactory->create();
                $quote = $quote->load($quoteID);
                $quote->setIsActive(false)->save();
                $customerBeforeAuthUrl = $this->url->getUrl('checkout');
                $this->responseFactory->create()->setRedirect($customerBeforeAuthUrl)->sendResponse();
                exit();
            }
        } catch (Throwable $t) {
            $this->logger->error($t->getMessage());
        }
    }
}`

**Comment by magent-harsh** [comment]
Hello 
we have facing same issue with after pay payment method get duplicate order with same product in magento 2.3.5
can you please give proper solution for that we have tried with all solution  above but no luck.

Thnak you

**Comment by trunglv** [comment]
This problem is so hard to detect a cause. This one https://github.com/magento/magento2/issues/25222 is one of a cause to make multiple duplicated orders. And Amazon Pay as well, I can't recall exactly.    
But generally, for somewhat reason is that the function Magento\Quote\Model\QuoteManagement::submitQuote is called on twice or even more.
- So my solution is to make a hook/plugin on the function "submitQuote" to manage a problem when it will happen - also we do a trace to log to know what happens. [ See code below]

So then we usually check a folder /var/order_duplicated_log. If there is a file with the prefix is "duplicated_", then a duplicated order issue would happen.  Base on a trace on a duplicated log file and logs from system.log of Magento, We can catch what code lines/3rd-party extension causes the issue.

I'm available if someone would want to hire me to fix it. Drop me an email to luuvantrung@gmail.com

I make a plugin : 
<?xml version="1.0"?>

```
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:ObjectManager/etc/config.xsd">
    <type name="Magento\Quote\Model\QuoteManagement">
        <plugin name="prevent-placeorder-exception" type="FixBugs\DuplicatedOrder\Plugin\QuoteManagement\SubmitQuote" sortOrder="1" />
    </type>

</config>
```

```
<?php
namespace FixBugs\DuplicatedOrder\Plugin\QuoteManagement;
use Magento\Framework\Event\Observer;
use Magento\Framework\Event\ObserverInterface;
use Magento\Framework\Message\ManagerInterface;
use Magento\Framework\App\RequestInterface;

use Magento\Sales\Model\Order\Email\Sender\OrderSender;


class SubmitQuote {
    
    public function __construct(
        \Magento\Framework\App\ResponseFactory $responseFactory,
        \Magento\Framework\UrlInterface $url,
        \Magento\Sales\Model\ResourceModel\Order\CollectionFactory $orderCollectionFactory,
        RequestInterface $request    
    ) {
        $this->responseFactory = $responseFactory;
        $this->url = $url;
        $this->orderCollectionFactory = $orderCollectionFactory;
        $this->request = $request;
    }

    public function beforeSubmit($subject,  $quote, $orderData = [] ) {
                
        $objectManager = \Magento\Framework\App\ObjectManager::getInstance();
        /** @var  $session \Magento\Framework\Session\Generic */
        $session   = $objectManager->get('Magento\Framework\Session\Generic');
        $response  = $objectManager->get('Magento\Framework\App\Response\Http');

        /** @var  $messageManager \Magento\Framework\Message\ManagerInterface */
        $messageManager = $objectManager->get('Magento\Framework\Message\ManagerInterface');

        /** @var  $chekcoutSession \Magento\Checkout\Model\Session */
        $chekcoutSession = $objectManager->get('Magento\Checkout\Model\Session');
        
        $orderSender = $objectManager->get('Magento\Sales\Model\Order\Email\Sender\OrderSender');
        
        if ( $quote &&
            $quote->getId() > 0 &&
            $quote->hasItems() &&
            $quote->getHasError() === null &&
            ($payment = $quote->getPayment()) &&            
            ($order = $chekcoutSession->getLastRealOrder()) &&
            $order->getStatus() === "processing"
            ) 
                {
                    //log information here - /var/order_duplicated_log - so we can check later - maybe find some useful information to fix this bug.
                    $relativeFileName = "duplicated_".$order->getIncrementId().'_'.time().'.log';
                    
                    $contents = array(
                        'time' => date("Y-m-d H:i:s"),                        
                        'request_info' => $_SERVER,
                        'params' => $this->request->getParams(),
                        'debugs' => $this->getBackTraces()
                            
                    );
                    $this->writeLogFile($relativeFileName, json_encode($contents, JSON_PRETTY_PRINT) );
                    
                    /*Send email if it has not been sent*/
                    if($order->getCanSendNewEmailFlag() && $order->getEmailSent()){
                        $orderSender->send($order);
                    }

                    /*Redirect to cart page -  Maybe here we can set void is inactive */

                    $session->destroy(['send_expire_cookie' => true]);
                    $messageManager->addNoticeMessage(
                        __('Message to a customer')
                    );
                    $response->setRedirect( "checkout/cart" )->sendResponse();
                    
                    die();
                    
        } else {
                  
                $relativeFileName = 'not_duplicated'.$quote->getId().'_'.time().'.log';    
                $contents = array(
                    'time' => date("Y-m-d H:i:s"),                        
                    'request_info' => $_SERVER,
                    'params' => $this->request->getParams(),
                    'debugs' => $this->getBackTraces()

                );
                $this->writeLogFile($relativeFileName, json_encode($contents, JSON_PRETTY_PRINT) );  
        }
        return [$quote, $orderData];
    }
    
    protected  function writeLogFile( $relativeFileName, $content ) {
        
        $om = \Magento\Framework\App\ObjectManager::getInstance();        
        $directory = $om->get('Magento\Framework\Filesystem\DirectoryList');
        $rootPath  =  $directory->getRoot();        
        $logFolder = $rootPath . DIRECTORY_SEPARATOR . 'var' . DIRECTORY_SEPARATOR . 'order_duplicated_log_plugin';
        
        if( !file_exists($logFolder) )  mkdir( $logFolder, 0777 );
        $relativeFileName = DIRECTORY_SEPARATOR . 'order_duplicated_log' . DIRECTORY_SEPARATOR . $relativeFileName;
        
        /** @var \Magento\Framework\Filesystem $filesystem */
        $filesystem = $om->get('Magento\Framework\Filesystem');
        
        /** @var \Magento\Framework\Filesystem\Directory\WriteInterface|\Magento\Framework\Filesystem\Directory\Write $writer */
        $writer = $filesystem->getDirectoryWrite('var');
        /** @var \Magento\Framework\Filesystem\File\WriteInterface|\Magento\Framework\Filesystem\File\Write $file */
        $file = $writer->openFile($relativeFileName, 'w');
        try {
            $file->lock();
            try {                
                $file->write($content);                
            } finally {                
                $file->unlock();                
            }
        } finally {
            $file->close();
        }
    }
    
    protected function getBackTraces(){
        
        $debugs = debug_backtrace();
        $debugClasses = array();
        foreach ($debugs as $debug) {
            $debugLines = array();

            $debugLines['file'] = isset($debug['file']) ? $debug['file'] : '';
            $debugLines['class'] = isset($debug['class']) ? $debug['class'] : '';
            $debugLines['method'] = isset($debug['function']) ? $debug['function'] : '';
            $debugLines['line'] = isset($debug['line']) ? $debug['line'] : '';

            //$debugClasses[] = implode("||", $debugLines);
            $debugClasses[] = $debugLines;
        }
        return $debugClasses;
    }
    
}
```

**Comment by magento-quality-patches** [comment]
The patch solving a similar issue is available in [Magento Quality Patches](https://support.magento.com/hc/en-us/articles/360047139492) package (MQP)
 
**Patch**
MDVA-31006: Fixes the issue where duplicated orders appear after placing an order using Paypal Express payment.
 
**Compatible versions**
Magento OpenSource/Commerce/Commerce Cloud 2.3.2 - 2.4.1
 
**Applying a patch - Magento OpenSource/Commerce**
 
1. `$ composer require magento/quality-patches`
2. `$ ./vendor/bin/magento-patches apply MDVA-31006`
 
See [MQP Magento Commerce documentation](https://devdocs.magento.com/guides/v2.4/comp-mgr/patching/mqp.html)
 
**Applying a patch - Magento Commerce Cloud**
See [MQP Magento Commerce Cloud documentation](https://devdocs.magento.com/cloud/project/project-patch.html)
 
**Patch Details**
https://support.magento.com/hc/en-us/articles/360050572151

**Comment by rifki** [comment]
Hi @magento-engcom-team Any update on this?

**Comment by JithinJay** [comment]
Hi can someone please help me with the file changes in patch: MDVA-31006 or any fixes applicable to Magento EE 2.3.1, I tried both the plugin and observer workarounds and the issue still not fixed.

**Comment by magento-quality-patches** [comment]
@JithinJay the patch MDVA-31006 was updated to be compatible with Magento 2.3.1. Feel free to try it using  [Magento Quality Patches](https://support.magento.com/hc/en-us/articles/360047139492) package (MQP)

**Comment by JancyAbraham** [comment]
Hi,
Can you please help me with the file changes in patch: MDVA-31006 or any fixes applicable to Magento Opensource 2.2.8.

**Comment by dufrygrant** [comment]
It's waiting on release of this commit:

https://github.com/Adyen/adyen-magento2/commit/a88db6a4fced6ff500fb0887cd0eca60da8028f6

**Comment by bartoszkubicki** [comment]
It's waiting on release of this commit:

Adyen/adyen-magento2@a88db6a

**Comment by shibinvr3** [comment]
Issue still exists with a free payment method in Magento 2.4.2. Any patch is available for this version?

**Comment by dchaykas** [comment]
Issue still exists in 2.4.1. The customer was able to place 4 orders using the same quote ID, the first 3 orders were created less than a minute apart. A slow response from the payment provider was flagged.

**Comment by willboudle** [comment]
So I see above, @rauberdaniel suggested making `quote_id ` unique and @DependencyHell said we cannot because there are instances where a quote id can be re-used, given the circumstance where an order is cancelled due to a failure and a new order is generated based on the quote. I suspect 'editing' an order may do the same thing. 

However, what about creating a UNIQUE constant on `state `and `quote_id`?  There really should never be a situation where that is duplicated right?

`ALTER TABLE `sales_order` ADD UNIQUE `orderstate_quoteid`(`state`, `quote_id`);`

**Comment by engcom-Alfa** [comment]
Hi @amitvishvakarma ,
As I have tried in 2.4-develop environment, I am not seeing the duplicated Quote_id.
Well I am referring to the **sales_order_item table with quote_item_id** column. **Kindly let me know if it is from specific table name.** I have used check/money order payment option and hope the payment method should not be dependent for this issue.
So, kindly check the below screen record that I tried to reproduce.

https://user-images.githubusercontent.com/51679138/121528779-699fb700-ca19-11eb-9dc7-50ed6375ef01.mp4

**Comment by engcom-Alfa** [comment]
As there is no response to my last comment in last ~14 days, closing this current issue.

Kindly upgrade to the latest version to re test. In case of any Magento related issues, you may please report a new bug with all the information like Preconditions, detailed steps to reproduce, expected and actual Results along with screenshots/ screen recordings possibly.

For any Magento related references, you may refer to the [Magento user guide](https://docs.magento.com/user-guide/).

**Comment by engcom-Hotel** [comment]
Hi @amitvishvakarma ,

Thanks for the suggestion!

Earlier we have tried it manually and unable to reproduce. We will try to reproduce the issue using some performance tools like JMeter. 

Thanks

**Comment by willboudle** [comment]
> Hi @amitvishvakarma ,
> As I have tried in 2.4-develop environment, I am not seeing the duplicated Quote_id.
> Well I am referring to the **sales_order_item table with quote_item_id** column. **Kindly let me know if it is from specific table name.** I have used check/money order payment option and hope the payment method should not be dependent for this issue.
> So, kindly check the below screen record that I tried to reproduce.
> 
>  2021-06-10.18-09-56.mp4 
> Kindly try to reproduce in the latest version and update me with complete reproducible steps and asked information again.

No one has answered this, but I'm curious if this will work.
I guess the only issue could be if we had to cancel more than one of the same order, there _could _ be two orders that are cancelled and share a quote_id.

**Comment by dchaykas** [comment]
I was able to replicate the issue with the quote ID being duplicated as many times as you trigger the "place order" CTA. I haven't had the time to submit a new issue but this one here is as close as it gets.

This was tested on 3 different Magento configurations.

**Setup used:**
Docker container
Magento 2.4.1
PHP 7.4

**Steps to reproduce:**
1. Add a product to the cart.
2. Go to checkout
3. Go through the checkout steps up until the last action of "Place order"
4. Navigate to your Docker instance and shutdown ElasticSearch
5. Press "Place Order"
6. Continue pressing "Place Order"

Expected Result:
I'm not sure what the exact flow in this case should be but given that the entire process is wrapped in a try/block and is calling for a rollback I'm thinking the order should not be created.

**Actual Result:**
The order is created, payment collected and the only thing that's causing this to stop is that Inventory cannot be updated because ES is down. The user is not redirected anywhere, just sits on the same page with a flash message at the top.
If you continue to press "Place Order" you'll continue to generate orders with the same quote ID.

**Comment by robertbaum** [comment]
Hello all,

I faced this issue recently and I don't know why now!
but while tracing the issue I found these notes

1. - happened for now with Twint Payment and offline payments too
2. - the address id for the payment and shipment id incremented 
3. - one of the orders coming from::1 and the other from::127.0.0.1

I think point 3 could help us to diagnose the issue 

![image](https://user-images.githubusercontent.com/46323639/132197471-e87dd706-1366-4c29-bd2d-4ee40a50ad1e.png)

**Comment by engcom-Delta** [comment]
Hi @willboudle,
Tried to reproduce the issue but in second order getting error like "Current Customer does not have active cart". Hence we are unable to place second order .

Added the label "Needs Update
<img width="960" alt="PlaceOrder_2_error" src="https://user-images.githubusercontent.com/51681379/134354421-cc42d215-1017-4298-8cf6-0feabb0d632f.png">
"

**Comment by wgabka** [comment]
Hello all. I have notice issue with duplicate orders with the same quote id. All of your cases can have different origin but it's generally related to bad architecture of exception handling and not using database transaction or locking mechanizm in \Magento\Quote\Model\QuoteManagement::submitQuote.

Problem is in this code fragment. 

```
        try {
            $order = $this->orderManagement->place($order);
            
            $quote->setIsActive(false);
            $this->eventManager->dispatch(
                'sales_model_service_quote_submit_success',
                [
                    'order' => $order,
                    'quote' => $quote
                ]
            );
            $this->quoteRepository->save($quote);
        } catch (\Exception $e) {
            $this->rollbackAddresses($quote, $order, $e);
            throw $e;
        }
``` 
author don't take in consideration that something can go wrong after line

```php
$order = $this->orderManagement->place($order);
```
and this is huge mistake because we have event there and magento can have a ton of plugins on place after.
What is happen if something goes wrong? Quote is not save and marked as not active. That means on frontend we will have the same quote active and users can try to create another order with the same quote id.

**Comment by fiko** [comment]
Hello All,

I tried to fix the issue, so I setIsActive to false once customer placing the order. And in order to handle (in case there is an error while placing order), I set active to true again.

Hope it's for good, I've test it at the same time almost 30 times. and it works well.

**Comment by engcom-Alfa** [comment]
This ticket had an internal Jira, and due to various reasons that Jira got processed and cancelled.
In order to further follow-up and take it forward, we have created a new issue - [35833](https://github.com/magento/magento2/issues/35833), and it has already created PR [35139](https://github.com/magento/magento2/pull/35139) linked.
Hence, we can refer these tickets for future updates.

Thanks for your all of your contributions and collaborations here!
